dags/veda_data_pipeline/veda_dataset_pipeline.py

```diff
@@ -83,19 +64,38 @@ def build_stac_task(payload):
 }
 
 with DAG("veda_dataset_pipeline", params=template_dag_run_conf, **dag_args) as dag:
-    # ECS dependency variable
+    start = EmptyOperator(task_id="start")
+    end = EmptyOperator(task_id="end")
 
-    start = EmptyOperator(task_id="start", dag=dag)
-    end = EmptyOperator(task_id="end", dag=dag)
 
-    collection_grp = collection_task_group()
-    discover = discover_from_s3_task.expand(event=extract_discovery_items())
-    discover.set_upstream(collection_grp)  # do not discover until collection exists
-    get_files = get_dataset_files_to_process(payload=discover)
+    @task()
+    def mutate_payload(**kwargs):
+        ti = kwargs.get("ti")
+        payload = ti.dag_run.conf.copy()
+        payloads = list()
+        if assets := payload.get("assets"):
+            # remove thumbnail asset if provided in collection config
+            if "thumbnail" in assets.keys():
+                assets.pop("thumbnail")
+            # if thumbnail was only asset, delete assets
+            if not assets:
+                payload.pop("assets")
+            # finally put the mutated assets back in the payload
+            else:
+                payload["assets"] = assets
+        for item in payload.get("discovery_items"):
+            payloads.append({
+                **payload,
+                **item
+            }
+            )
+
+        return payloads
 
-    build_stac = build_stac_task.expand(payload=get_files)
-    # .output is needed coming from a non-taskflow operator
-    submit_stac = submit_to_stac_ingestor_task.expand(built_stac=build_stac)
 
-    collection_grp.set_upstream(start)
-    submit_stac.set_downstream(end)
+    mutated_payloads = start >> collection_task_group() >> mutate_payload()
+    run_discover_build_and_push = TriggerDagRunOperator.partial(
+        task_id="trigger_discover_items_dag",
+        trigger_dag_id="veda_discover",
+        wait_for_completion=True
+    ).expand(conf=mutated_payloads) >> end
```

Comment:
I get that this promotes reuseability at the level of the `discover` DAG, but I worry that this creates too much of a breakdown in reuse and observability at the task level.

- Rather than passing task-specific data structures as parameters, we end up relying on `ti.dag_run.conf`. This means that, unless we always build and ingest items through the discover DAG, we cannot rely on the data format being the same for different DAGs with different ingestion strategies. This will eventually bring us back to the same solution we have now, where new DAGs skip using the `discover` DAG, and simply reuse the tasks with additional wrapper steps to manipulate the input. We should try to lean into this - the changes you made to `get_files()` is a great example, where we promote modularity and reuse at the task level, so that it doesn't matter what the incoming `dag_run.conf` is.
- Similarly, this condenses task status into a single node in the `dataset` DAG. Meanwhile, in the triggered `discover` DAG, there are several expanded steps with unique failure conditions. This means that a failure in one step in one triggered DAG will require a retry of the complete DAG, rather than a single task.
- `TriggerDagRunOperator` adds a link between DAGs, but not between executions. I was hoping this would work better with `expand()`, but unfortunately we're out of luck. This (and a few similar issues) is being [tracked as a bug ](https://github.com/apache/airflow/issues/41977) in Airflow, but until it's fixed, I think we should steer clear so as to maintain execution-level observability, especially in DAG runs with a large number of mapped discovery tasks.

Reply:
Yes, I agree that traceability will be necessary for debugging any issues. Some have reported that this problem is fixed in Airflow 2.10. However, we still need to promote task reusability to avoid defining the same logic twice. I will close this PR and open a new one with some refactoring to make debugging and development easier.

Reply:
Closing this in favor of https://github.com/NASA-IMPACT/veda-data-airflow/pull/268